janitor/utils.py

```diff
@@ -405,3 +406,219 @@ def _wrapped(x, *args, **kwargs):
+                raise ValueError("passed DataFrame cannot be empty")
+            if not (
+                isinstance(value.index, pd.MultiIndex)
+                or isinstance(value.columns, pd.MultiIndex)
+            ):
+                # add key to dataframe columns
+                value = value.add_prefix(f"{key}_")
+                dfs.append(value)
+            else:
+                raise ValueError(
+                    "expand grid does not work with pd.MultiIndex"
+                )
+        # process lists
+        if isinstance(value, list):
+            if not value:
+                raise ValueError("passed value cannot be empty")
+            elif np.array(value).ndim == 1:
+                checklist = (type(None), str, int, float, bool)
+                check = (isinstance(internal, checklist) for internal in value)
+                if all(check):
+                    dicts.update({key: value})
+                else:
+                    raise ValueError("values in iterable must be scalar")
+            elif np.array(value).ndim == 2:
+                value = pd.DataFrame(value).add_prefix(f"{key}_")
+                dfs.append(value)
+            else:
+                raise ValueError("sequence's dimension should be 1d or 2d")
+
+    return dfs, dicts
+
+
+def _grid_computation_dict(dicts: Dict):
+    """
+    Function used within the expand_grid function,
+    to compute dataframe from values that are not dataframes/arrays/series.
+    These values are collected into a dictionary,
+    and processed with numpy meshgrid. Numpy's meshgrid is faster than itertools' product,
+    and when converting to a dataframe, is fast as well.
+    Structured arrays are used here, to ensure the datatypes are preserved.
+    """
+    # if there is only name value pair in the dictionary
+    if len(dicts) == 1:
+        key = list(dicts.keys())[0]
+        value = list(dicts.values())[0]
+        final = pd.DataFrame(value, columns=[key])
+    # if there are more than one name value pair
+    else:
+        res = np.meshgrid(*dicts.values())
+        # create structured array
+        # keeps data type of each value in the dict
+        outcome = np.core.records.fromarrays(res, names=",".join(dicts))
+        # reshape into a 1 column array
+        # using the size of any of the arrays obtained from the meshgrid computation
+        outcome = np.reshape(outcome, (np.size(res[0]), 1))
+        # flatten structured array into 1d array
+        outcome = np.concatenate(outcome)
+        # sort array
+        outcome.sort(axis=0, order=list(dicts))
+        # create dataframe
+        # structed array already has names,
+        # this gets transferred as column names
+        final = pd.DataFrame.from_records(outcome)
+    return final
+
+
+# this is for dataframes/series
+# this should be the final output if there are lists or lists and dicts
+# returned from check instance
+def _compute_two_dfs(df1: pd.DataFrame, df2: pd.DataFrame):
```

Comment:
Return type annotation needed:

```suggestion
def _compute_two_dfs(df1: pd.DataFrame, df2: pd.DataFrame) -> pd.DataFrame:
```